src/mip/extras/recording_connection.hpp

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <mip/mip_device.hpp>
+
+#include <memory>
+#include <ostream>
+#include <iostream>
+
+namespace mip
+{
+namespace extras
+{
+
+////////////////////////////////////////////////////////////////////////////////
+///@addtogroup mip_extras Extra utilities
+///@{
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Can be used with another connection to communicate with a device, and record the data at the same time
+///
+class RecordingConnection : public Connection
+{
+public:
+    RecordingConnection(Connection* connection, std::ostream* recvFile=nullptr, std::ostream* sendFile=nullptr);
+
+    bool sendToDevice(const uint8_t* data, size_t length) final;
+    bool recvFromDevice(uint8_t* buffer, size_t max_length, size_t* count_out, Timestamp* timestamp_out) final;
+
+protected:
+    Connection* mConnection;
+
+    // Files may be NULL to not record one direction or the other
+    std::ostream* mRecvFile;
+    std::ostream* mSendFile;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Template wrapper for a recording connection.
+///
+///@param ConnectionType The type of connection used to actually communicate.
+///
+template<typename ConnectionType>
+class RecordingConnectionWrapper : public RecordingConnection
+{
+public:
+    ///@brief Creates a RecordingConnectionWrapper that will write received bytes to recvFile, sent bytes to sendFile, and construct a connection object from args
+    ///
+    ///@param recvFile The file to write to when bytes are received. Null if received bytes should not be written to a file
+    ///@param sendFile The file to write to when bytes are sent. Null if sent bytes should not be written to a file
+    ///@param args     Arguments required to construct the ConnectionType
+    template<class... Args>
+    RecordingConnectionWrapper(std::ostream* recvFile, std::ostream* sendFile, Args&&... args) :
+        mConnectionPtr(std::unique_ptr<ConnectionType>(new ConnectionType(std::forward<Args>(args)...))), RecordingConnection(nullptr, recvFile, sendFile)
+    {
+        mConnection = mConnectionPtr.get();
+    }
+private:
+    std::unique_ptr<ConnectionType> mConnectionPtr;
```

Comment:
Technically the pointer doesn't need to be stored twice (once here and once in the base class). Instead you could make the destructor free the pointer(s?) from the base class. That said, the intent of this is pretty clear and it's pretty robust against future changes breaking it.
Would there be a benefit to keeping the ostreams in here as well?

Reply:
That is a good point I did not think of. I think that not storing the pointer twice is the way to go.

I don't think there would be any benefit of doing the same for the streams as we would need to construct them in the constructor in order to free them. We could certainly do this, but the error handling could be messy, and I think I would rather have the user manage their own files